crates/python_extractor/src/lib.rs

```diff
@@ -0,0 +1,183 @@
+use std::{
+    collections::{BTreeSet, HashSet},
+    io::Read,
+    path::PathBuf,
+};
+
+use bzl_gen_build_python_utilities::PythonProgram;
+
+use bzl_gen_build_shared_types::api::extracted_data::{DataBlock, ExtractedData};
+
+use anyhow::{Context, Result};
+
+mod extract_py_bzl_gen_build_commands;
+mod extract_py_imports;
+
+pub async fn extract_python(
+    relative_input_paths: String,
+    working_directory: PathBuf,
+    output: PathBuf,
+    label_or_repo_path: String,
+    disable_ref_generation: bool,
+    import_path_relative_from: Option<String>,
+) -> Result<()> {
+    let mut relative_input_paths: Vec<String> =
+        if let Some(suffix) = relative_input_paths.strip_prefix('@') {
+            std::fs::read_to_string(PathBuf::from(suffix))?
+                .lines()
+                .map(|e| e.to_string())
+                .collect()
+        } else {
+            vec![relative_input_paths.clone()]
+        };
+
+    relative_input_paths.sort();
+
+    let mut data_blocks: Vec<DataBlock> = Default::default();
+
+    for relative_path in relative_input_paths {
+        let input_file = working_directory.join(&relative_path);
+        let mut refs: HashSet<String> = Default::default();
+        let mut defs: BTreeSet<String> = Default::default();
+        let mut bzl_gen_build_commands: HashSet<String> = Default::default();
+
+        let mut file = std::fs::File::open(&input_file).with_context(|| {
```

Comment:
From this line down to line 60 are the only real changes in here. Notably, it now uses `from_utf8_lossy` which will replace non-utf8 characters with �